### DISCUSSION TITLE
如何获取 Cascader 组件选择框的宽度

### DISCUSSION THREAD

**Discussion by Firstsup** [ant-design/ant-design#44510] [Q&A]
我在 Cascader 组件之上封装新组件，需要获取 Cascader 组件选择框的宽度，但是组件的 ref 中并没有包含，有什么方法可以拿到吗？

**Comment by Firstsup** [comment]
Cascader 的 ref 中只包含 blur、focus 和 scrollTo，拿不到组件宽度等属性。请各位大佬指导一下

**Reply by kiner-tang** [reply]
或许可以尝试参考这个示例，在 dropdownRender 里面去获取相关的信息：https://ant.design/components/cascader-cn#cascader-demo-custom-dropdown

**Reply by Firstsup** [reply]
在 dropdownRender 的 menus 里面找遍了，也没有找到 style 样式，好像不太行...

**Reply by kiner-tang** [reply]
你是想要获取面板宽度，还是想要给面板设置样式？如果获取面板宽度的话，在 dropdownRender 里面返回的根对象绑定上 ref 就可以了，如果要给面板设置样式，也可以直接用 dropdownMenuColumnStyle 看看
<img width="1316" alt="image" src="https://github.com/ant-design/ant-design/assets/10286961/98f5d296-496a-4228-bf37-37e81e120d21">
https://codesandbox.io/s/kuo-zhan-cai-dan-antd-5-8-5-forked-zls36f?file=/demo.tsx

**Reply by Firstsup** [reply]
非常感谢大佬的指导，但是我的有点不同，我的需求是需要在组件渲染时就获取到上方“选择框”的宽度。
经过您的思路的启发，我试了通过设置 dropdownMatchSelectWidth 使下拉面板和选择框同宽，然后给根组件绑定 ref，成功拿到了下拉面板的宽度，也间接获取到了选择框的宽度，但是问题是，我需要在第一时间就拿到选择框的宽度，但是这种方式必须要点开下拉面板，才能通过 ref 拿到宽度，所以也是不合适的。
十分感谢大佬百忙之中的指导，还想请问一下，如果想在第一时间就拿到“选择框”的宽度，有方法吗？

**Reply by kiner-tang** [reply]
看来我应该是理解错你的意思了，你是想要获取上面那个选择框的宽度，而不是下面的实际选择面板的宽度。目前 ref 没有将 dom 暴露出来，没办法获取。或许可以直接用 querySelector 的方式获取：
https://codesandbox.io/s/kuo-zhan-cai-dan-antd-5-8-5-forked-zls36f?file=/demo.tsx

**Reply by Firstsup** [reply]
这个方法考虑了一下也不合适...没关系，谢谢大佬的帮助，非常感谢！

**Comment by afc163** [comment]
最好描述一下原始诉求，避免 X-Y 问题。

**Reply by Firstsup** [reply]
谢谢大佬！我的需求是在 Cascader 组件之上封装组件，在 options 为单级数据的时候，最初要求下拉面板与选择框等宽，通过设置 dropdownMatchSelectWidth 属性进行实现，但是发现如果选项文字过长，文字会被直接截断。

为了解决这个问题，我想给选项通过配置成 Typography.Text 并加上 ellipsis 属性，让超出部分进行隐藏，但是使用 ellipsis 需要给出 width（即超过多宽才隐藏），因此希望拿到选择框的宽度，从而设置这个 width 的值。

或者另一种实现方案也可以：如果有选项内容宽度比选择框宽度长，那么下拉面板宽度就自适应（表现是比选择框长），否则保持和选择框等宽，这就需要设置下拉面板的 min-width 与选择框宽度相同，同样因为当前拿不到选择框的宽度而无法实现。

再次感谢您在百忙之中抽出时间回复我的问题！

**Reply by afc163** [reply]
当 options 为单级数据时，直接切换成 Select 如何呢？

**Reply by Firstsup** [reply]
十分感谢您的回复！我考虑过这种方案，当时否决的原因其一是 Cascader 和 Select 组件的 UI 差异较大，例如 Cascader 多选的时候是前面有选择框，而 Select 是没有选择框且后面有✅，差异较大；其二是之前已经在 Cascader 之上添加了很多新功能，如果要对于单级数据底层换用 Select，功能要重新开发，成本过高。

所以您的意思是，可能就是没有办法获取到 Cascader 组件的选择框的宽度，是吗？实在没办法也没关系的，我也有一些其他不完美的方法替代一下（例如数据单级时给下拉面板一个合理固定数值的最小宽度，虽然不能保证下拉面板和选择框同宽，但起码在大多数场景下组件表现能保持一致，且内容过长也会自适应，不会截断），也没有关系的。

**Reply by afc163** [reply]
不知道利用主题能力直接设置固定宽度是否能满足你的需求：https://codesandbox.io/s/ji-ben-antd-5-8-5-forked-9s8j95?file=/demo.tsx

**Reply by Firstsup** [reply]
谢谢您的方法，但是这样就丧失了灵活度，新组件作为一个基础组件，我还是不能给它限制固定宽度，因为组件会用在各种场景上，宽度也会有所不同，不能通用地给出一个固定的宽度。所以这种方法可能也不合适...

**Reply by afc163** [reply]
当 options 为单级数据时，`dropdownMatchSelectWidth` 设置为 true 呢。我才发现 dropdownMatchSelectWidth 也是有用的。

**Reply by Firstsup** [reply]
这个是我在第一个对您的回复中提到的问题，如果对于单级数据设置 dropdownMatchSelectWidth 为 true，如果选项文字长度过长（超过选择框的宽度），选项文字就会出现截断效果，例如这个 demo：https://codesandbox.io/s/ji-ben-antd-5-8-5-forked-6fkqmj?file=/demo.tsx

所以直接使用 dropdownMatchSelectWidth 是不行的。

**Reply by afc163** [reply]
这是个 bug，可以修掉。

**Reply by Firstsup** [reply]
嗯嗯，谢谢大佬的帮助和解答，感谢您！
